modules/favorites/module.php

```diff
@@ -0,0 +1,323 @@
+				$type_instance->filter(
+					function( $value ) use ( $favorites ) {
+						return ! in_array( $value, $favorites, true );
+					}
+				);
+				break;
+			default:
+				$this->action_doesnt_exists( $action );
+		}
+
+		if ( $store && ! $this->store() ) {
+			return false;
+		}
+
+		return $type_instance->values();
+	}
+
+	/**
+	 * Check whether favorites type is registered.
+	 *
+	 * @param string $type
+	 *
+	 * @return bool
+	 */
+	public function exists( $type ) {
+		return in_array( $type, array_keys( $this->types ), true );
+	}
+
+	/**
+	 * Get registered favorites type instance.
+	 *
+	 * @param string $type
+	 *
+	 * @return Favorites_Type
+	 */
+	public function type_instance( $type ) {
+		if ( ! $this->exists( $type ) ) {
+			$this->type_doesnt_exists( $type );
+		}
+
+		return $this->types[ $type ];
+	}
+
+	/**
+	 * Register a new type class.
+	 *
+	 * @param string $class
+	 */
+	public function register( $class ) {
+		if ( ! class_exists( $class ) ) {
+			$this->class_doesnt_exists( $class );
+		}
+
+		$type_instance = new $class();
+
+		$this->types[ $type_instance->get_name() ] = $type_instance;
+	}
+
+	/**
+	 * Returns all available types keys.
+	 *
+	 * @return string[]
+	 */
+	public function available() {
+		return array_keys( $this->types );
+	}
+
+	/**
+	 * Combine favorites from all types into a single array.
+	 *
+	 * @return array
+	 */
+	protected function combined() {
+		$all = [];
+
+		foreach ( $this->types as $type ) {
+			$favorites = $type->values();
+
+			if ( ! empty( $favorites ) ) {
+				$all[ $type->get_name() ] = $favorites;
+			}
+		}
+
+		return $all;
+	}
+
+	/**
+	 * Populate all type classes with the stored data.
+	 */
+	protected function populate() {
+		$combined = $this->retrieve();
+
+		foreach ( $this->types as $key => $type ) {
+			if ( isset( $combined[ $key ] ) ) {
+				$type->merge( $combined[ $key ] );
+			}
+		}
+	}
+
+	/**
+	 * Retrieve stored user favorites types.
+	 *
+	 * @return mixed|false
+	 */
+	protected function retrieve() {
+		return get_user_option( static::OPTION_NAME );
+	}
+
+	/**
+	 * Update all changes to user favorites type.
+	 *
+	 * @return int|bool
+	 */
+	protected function store() {
+		return update_user_option( get_current_user_id(), static::OPTION_NAME, $this->combined() );
+	}
+
+	/**
+	 * Display action doesn't exists error.
+	 *
+	 * @param string $action
+	 */
+	public function action_doesnt_exists( $action ) {
+		wp_die(
+			new WP_Error(
+				'action_doesnt_exists',
+				sprintf(
+					esc_html__( "Action '%s' to apply on favorites doesn't exists", 'elementor' ),
+					$action
+				)
+			)
+		);
+	}
+
+	/**
+	 * Display type doesn't exists error.
+	 *
+	 * @param string $type
+	 */
+	public function type_doesnt_exists( $type ) {
+		wp_die(
+			new WP_Error(
+				'type_doesnt_exists',
+				sprintf(
+					esc_html__( "Favorites type '%s' doesn't exists", 'elementor' ),
+					$type
+				)
+			)
+		);
+	}
+
+	/**
+	 * Display class doesn't exists error.
+	 *
```

Comment:
If the errors are for the development stage, you can leave them to fail via PHP Errors.